x-pack/plugins/endpoint/server/routes/resolver/utils/fetch.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IScopedClusterClient } from 'kibana/server';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { getPaginationParams } from './pagination';
+import { Tree } from './tree';
+import { LifecycleQuery } from '../queries/lifecycle';
+import { ChildrenQuery } from '../queries/children';
+import { EventsQuery } from '../queries/events';
+import { StatsQuery } from '../queries/stats';
+
+export class Fetcher {
+  constructor(
+    private readonly client: IScopedClusterClient,
+    private readonly id: string,
+    private readonly endpointID?: string
+  ) {}
+
+  public async ancestors(limit: number): Promise<Tree> {
+    const tree = new Tree(this.id);
+    await this.doAncestors(tree, this.id, this.id, limit);
+    return tree;
+  }
+
+  public async children(limit: number, generations: number, after?: string): Promise<Tree> {
```

Comment:
Actually all 4 of these functions can return lists of events with their pagination and then use the add* functions in Tree to operate on a single tree rather than making 4 separate trees and merging them together.  In addition to avoiding the tree merge operations, it also means the handleEvents function won't need to deal with the Tree structure at all.  This makes it clear that the response from that handler is really just a list and not a tree.